File: src/chronic_types/duration.rs
use std::fmt::Write;
// a positive duration
// a datetime type representing datetime points since unix epoch
// advance datetime by given duration
// go back datetime by given duration

pub const MS_PER_SECOND: u64 = 1000;
pub const MS_PER_MINUTE: u64 = MS_PER_SECOND * 60;
pub const MS_PER_HOUR: u64 = MS_PER_MINUTE * 60;
pub const MS_PER_DAY: u64 = MS_PER_HOUR * 24;
pub const MS_PER_WEEK: u64 = MS_PER_DAY * 7;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Duration(u64);

impl Duration {
  pub const ZERO: Duration = Duration(0);
  
  pub const fn from_milliseconds(milliseconds: u64) -> Self {
    Self(milliseconds)
  }

  pub const fn from_seconds(seconds: u64) -> Option<Self> {
    if let Some(milliseconds) = seconds.checked_mul(MS_PER_SECOND) {
      Some(Self(milliseconds))
    } else {
      None
    }
  }

  pub const fn from_minutes(minutes: u64) -> Option<Self> {
    if let Some(milliseconds) = minutes.checked_mul(MS_PER_MINUTE) {
      Some(Self(milliseconds))
    } else {
      None
    }
  }
  
  pub const fn unchecked_from_minutes(minutes: u64) -> Self {
    Self(minutes * MS_PER_MINUTE)
  }

  pub const fn unchecked_from_hours(hours: u64) -> Self {
    Self(hours * MS_PER_HOUR)
  }

  pub const fn from_hours(hours: u64) -> Option<Self> {
    if let Some(milliseconds) = hours.checked_mul(MS_PER_HOUR) {
      Some(Self(milliseconds))
    } else {
      None
    }
  }

  pub const fn from_days(days: u64) -> Option<Self> {
    if let Some(milliseconds) = days.checked_mul(MS_PER_DAY) {
      Some(Self(milliseconds))
    } else {
      None
    }
  }

  pub const fn from_weeks(weeks: u64) -> Option<Self> {
    if let Some(milliseconds) = weeks.checked_mul(MS_PER_WEEK) {
      Some(Self(milliseconds))
    } else {
      None
    }
  }

  pub const fn new_week() -> Self {
    Self(7 * MS_PER_DAY)
  }

  pub const fn unchecked_from_days(days: u64) -> Self {
    Self::from_milliseconds(days * MS_PER_DAY)
  }

  pub const fn unchecked_from_days_u32(days: u32) -> Self {
    Self::from_milliseconds(days as u64 * MS_PER_DAY)
  }

  pub const fn is_zero(&self) -> bool {
    self.0 == 0
  }

  pub const fn total_seconds(&self) -> u64 {
    self.0 / MS_PER_SECOND
  }

  pub const fn total_minutes(&self) -> u64 {
    self.0 / MS_PER_MINUTE
  }

  pub const fn total_hours(&self) -> u64 {
    self.0 / MS_PER_HOUR
  }

  pub const fn total_days(&self) -> u64 {
    self.0 / MS_PER_DAY
  }

  pub const fn total_weeks(&self) -> u64 {
    self.0 / MS_PER_WEEK
  }

  pub const fn total_milliseconds(&self) -> u64 {
    self.0
  }

  pub fn checked_add(&self, other: &Duration) -> Option<Self> {
    self.0.checked_add(other.0).map(Self::from_milliseconds)
  }

  pub fn unchecked_add(&self, other: &Duration) -> Self {
    Self(self.0 - other.0)
  }

  pub fn checked_sub(&self, other: &Duration) -> Option<Self> {
    self.0.checked_sub(other.0).map(Self::from_milliseconds)
  }

  pub fn unchecked_sub(&self, other: &Duration) -> Self {
    Self(self.0 - other.0)
  }

  pub fn checked_mul(&self, by: u64) -> Option<Self> {
    self.0.checked_mul(by).map(Self::from_milliseconds)
  }

  pub fn checked_div(&self, by: u64) -> Option<Self> {
    self.0.checked_div(by).map(Self::from_milliseconds)
  }

  pub fn to_std(&self) -> std::time::Duration {
    std::time::Duration::from_millis(self.0)
  }

  pub fn to_chrono(&self) -> Option<chrono::Duration> {
    Some(chrono::Duration::milliseconds(match self.0.try_into() {
      Ok(value) => value,
      Err(_) => return None,
    }))
  }

  pub fn to_string(&self) -> String {
    let mut string = String::new();
    let mut milliseconds = self.total_milliseconds();

    let days = milliseconds / MS_PER_DAY;
    milliseconds %= MS_PER_DAY;
    
    let hours = milliseconds / MS_PER_HOUR;
    milliseconds %= MS_PER_HOUR;

    let minutes = milliseconds / MS_PER_MINUTE;
    milliseconds %= MS_PER_MINUTE;

    let seconds = milliseconds / MS_PER_SECOND;
    // milliseconds %= MS_PER_SECOND;

    if days > 0 {
      write!(string, "{}D", days).unwrap();
    }
    if hours > 0 {
      if string.len() > 0 {
        write!(string, " {}H", hours).unwrap();
      } else {
        write!(string, "{}H", hours).unwrap();
      }
    }
    if minutes > 0 {
      if string.len() > 0 {
        write!(string, " {}M", minutes).unwrap();
      } else {
        write!(string, "{}M", minutes).unwrap();
      }
    }
    if seconds > 0 {
      if string.len() > 0 {
        write!(string, " {}S", seconds).unwrap();
      } else {
        write!(string, "{}S", seconds).unwrap();
      }
    }

    string
  }
}

use serde::{Serialize, Serializer, Deserialize, Deserializer};

impl Serialize for Duration {
  fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
  where
    S: Serializer,
  {
    serializer.serialize_u64(self.0)
  }
}

impl<'de> Deserialize<'de> for Duration {
  fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
  where
    D: Deserializer<'de>,
  {
    Ok(Duration::from_milliseconds(u64::deserialize(deserializer)?))
  }
}

pub mod database_serde {
  use crate::database::*;
  use crate::GenericError;
  use super::Duration;

  // pub struct Adapter;

  // impl Adapter {
  //   pub fn new() -> Self {
  //     Self {}
  //   }
  // }

  // impl ScalarTypeAdapter for Adapter {
  //   type Type = Duration;

  //   fn serialize(&self, value: &Self::Type, context: SerializeScalarValueContext) {
  //     context.write_u64(self.total_milliseconds());      
  //   }

  //   fn deserialize(&self, value: ColumnValue) -> Result<Self::Type, GenericError> {
  //     let milliseconds = value.as_u64().map_err(|error|
  //       error.change_context("Failed to create a Duration from a ColumnValue: ColumnValue is not a u64 number")
  //     )?;

  //     Ok(Duration::from_milliseconds(value))        
  //   }
  // }

  impl SerializableScalarValue for Duration {
    fn write_into(&self, context: &mut SerializeScalarValueContext) -> Result<(), GenericError> {
      context.write_u64(self.total_milliseconds())
    }
  }

  impl DeserializableScalarValue for Duration {
    fn deserialize(value: ScalarValue) -> Result<Self, GenericError> {
      value
        .as_u64()
        .map(Duration::from_milliseconds)
        .map_err(|error|
          error.change_context("deserializing a duration")
        )
    }
  }
}

File: discipline_daemon_simple_ts/Lib/Time/Duration.ts
import { None, Option, Some } from "../../Prelude.ts";
import { Units } from "./mod.ts"

export type Duration = number & { readonly __type: "Time.Duration" }

export const Duration = {
  newZero(): Duration {
    return 0 as Duration
  },

  newOneMinute(): Duration {
    return Units.millisecondsPerMinute as Duration
  },

  new2Minutes(): Duration {
    return Units.millisecondsPerMinute * 2 as Duration
  },

  newWeek(): Duration {
    return Units.millisecondsPerWeek as Duration
  },
  
  uncheckedFromMilliseconds(milliseconds: number): Duration {
    if (Number.isSafeInteger(milliseconds) && milliseconds >= 0) {
      return milliseconds as Duration
    }
    throw new Error("Invaid argument: " + milliseconds.toString())
  },

  fromMilliseconds(milliseconds: number): Option<Duration> {
    if (Number.isSafeInteger(milliseconds) && milliseconds >= 0) {
      return Some(milliseconds as Duration)
    } else {
      return None()
    }
  },

  fromSeconds(seconds: number): Option<Duration> {
    if (Number.isSafeInteger(seconds) && seconds >= 0) {
      const millis = seconds * Units.millisecondsPerSecond
      if (millis === Infinity) {
        return None()
      } else {
        return Some(millis as Duration)
      }
    }

    return None()
  },

  fromMinutes(minutes: number): Option<Duration> {
    if (Number.isSafeInteger(minutes) && minutes >= 0) {
      const millis = minutes * Units.millisecondsPerMinute
      if (millis === Infinity) {
        return None()
      } else {
        return Some(millis as Duration)
      }
    }

    return None()
  },

  fromHours(hours: number): Option<Duration> {
    if (Number.isSafeInteger(hours) && hours >= 0) {
      const millis = hours * Units.millisecondsPerHour
      if (millis === Infinity) {
        return None()
      } else {
        return Some(millis as Duration)
      }
    }

    return None()
  },
  
  fromDays(days: number): Option<Duration> {
    if (Number.isSafeInteger(days) && days >= 0) {
      const millis = days * Units.millisecondsPerDay
      if (millis === Infinity) {
        return None()
      } else {
        return Some(millis as Duration)
      }
    }

    return None()
  },
  
  fromWeeks(weeks: number): Option<Duration> {
    if (Number.isSafeInteger(weeks) && weeks >= 0) {
      const millis = weeks * Units.millisecondsPerWeek
      if (millis === Infinity) {
        return None()
      } else {
        return Some(millis as Duration)
      }
    }

    return None()
  },

  milliseconds(me: Duration): number {
    return me
  },

  seconds(me: Duration): number {
    return Units.millisecondsToSeconds(me)
  },

  minutes(me: Duration): number {
    return Units.millisecondsToMinutes(me)
  },

  hours(me: Duration): number {
    return Units.millisecondsToHours(me)
  },

  days(me: Duration): number {
    return Units.millisecondsToDays(me)
  },

  weeks(me: Duration): number {
    return Units.millisecondsToWeeks(me)
  },

  eq(me: Duration, other: Duration) {
    return me === other
  },

  gth(a: Duration, b: Duration) {
    return a > b
  },

  gte(a: Duration, b: Duration) {
    return a >= b
  },

  lth(a: Duration, b: Duration) {
    return a < b
  },

  lte(a: Duration, b: Duration) {
    return a <= b
  },

  isZero(me: Duration) {
    return me === 0
  },

  clone(me: Duration): Duration {
    return me
  },

  sub(lhs: Duration, rhs: Duration): Option<Duration> {
    if (lhs >= rhs) {
      return Some(lhs - rhs as Duration)
    } else {
      return None()
    }
  },

  subOrZero(lhs: Duration, rhs: Duration) {
    if (lhs <= rhs) {
      return Duration.newZero()
    } else {
      return lhs - rhs as Duration
    }
  },

  add(lhs: Duration, rhs: Duration): Option<Duration> {
    const result = lhs + rhs
    if (Number.isSafeInteger(result) && result >= 0) {
      return Some(Duration.uncheckedFromMilliseconds(result))
    } else {
      return None()
    }
  },

  toPrettyFormat(me: Duration): string {
    let milliseconds = Duration.milliseconds(me)

    const days = Math.floor(milliseconds / Units.millisecondsPerDay);
    milliseconds %= Units.millisecondsPerDay;

    const hours = Math.floor(milliseconds / Units.millisecondsPerHour);
    milliseconds %= Units.millisecondsPerHour;

    const minutes = Math.floor(milliseconds / Units.millisecondsPerMinute);
    milliseconds %= Units.millisecondsPerMinute;

    const seconds = Math.floor(milliseconds / Units.millisecondsPerSecond);

    let string = ""
    function push(value: string) {
      if (string.length) {
        string += " "
      }
      string += value
    }

    if (days > 0) {
      push(`${days}D`);
    }
    if (hours > 0) {
      push(`${hours}H`);
    }
    if (minutes > 0) {
      push(`${minutes}M`);
    }
    if (seconds > 0) {
      push(`${seconds}S`);
    }

    return string.length > 0 ? string : 'Zero';
  },

  min(a: Duration, b: Duration): Duration {
    return a < b ? a : b
  }
}

export type MaybeDuration = Option<Duration>

export const MaybeDuration = {
  eq(a: MaybeDuration, b: MaybeDuration) {
    return Option.eq(a, b, Duration.eq)
  }
}